pkg/api/controllers/attachment.go

```diff
@@ -0,0 +1,359 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+This module implements a entry into the OpenSDS northbound service.
+
+*/
+
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	log "github.com/golang/glog"
+	"github.com/opensds/opensds/contrib/drivers/utils/config"
+	"github.com/opensds/opensds/pkg/api/policy"
+	c "github.com/opensds/opensds/pkg/context"
+	"github.com/opensds/opensds/pkg/controller/client"
+	"github.com/opensds/opensds/pkg/db"
+	"github.com/opensds/opensds/pkg/model"
+	pb "github.com/opensds/opensds/pkg/model/proto"
+	"github.com/opensds/opensds/pkg/utils"
+	apiconfig "github.com/opensds/opensds/pkg/utils/config"
+)
+
+func NewVolumeAttachmentPortal() *VolumeAttachmentPortal {
+	return &VolumeAttachmentPortal{
+		CtrClient: client.NewClient(),
+	}
+}
+
+type VolumeAttachmentPortal struct {
+	BasePortal
+
+	CtrClient client.Client
+}
+
+func (v *VolumeAttachmentPortal) CreateVolumeAttachment() {
+	if !policy.Authorize(v.Ctx, "volume:create_attachment") {
+		return
+	}
+	ctx := c.GetContext(v.Ctx)
+	var attachment = model.VolumeAttachmentSpec{
+		BaseModel: &model.BaseModel{},
+	}
+
+	if err := json.NewDecoder(v.Ctx.Request.Body).Decode(&attachment); err != nil {
+		errMsg := fmt.Sprintf("parse volume attachment request body failed: %s", err.Error())
+		v.ErrorHandle(model.ErrorBadRequest, errMsg)
+		return
+	}
+
+	// Check if host exists
+	host, err := db.C.GetHost(ctx, attachment.HostId)
+	if err != nil {
+		errMsg := fmt.Sprintf("get host failed in create volume attachment method: %v", err)
+		v.ErrorHandle(model.ErrorBadRequest, errMsg)
+		return
+	}
+
+	// Check if volume exists and volume status is normal
+	vol, err := db.C.GetVolume(ctx, attachment.VolumeId)
+	if err != nil {
+		errMsg := fmt.Sprintf("get volume failed in create volume attachment method: %v", err)
+		v.ErrorHandle(model.ErrorBadRequest, errMsg)
+		return
+	}
+
+	if vol.Status == model.VolumeAvailable {
+		db.UpdateVolumeStatus(ctx, db.C, vol.Id, model.VolumeAttaching)
+	} else if vol.Status == model.VolumeInUse {
+		if vol.MultiAttach {
+			db.UpdateVolumeStatus(ctx, db.C, vol.Id, model.VolumeAttaching)
+		} else {
+			errMsg := "volume is already attached to one of the host. If you want to attach to multiple host, volume multiattach must be true"
+			v.ErrorHandle(model.ErrorBadRequest, errMsg)
+			return
+		}
+	} else {
+		errMsg := "status of volume is available. It can be attached to host"
+		v.ErrorHandle(model.ErrorBadRequest, errMsg)
+		return
+	}
+
+	// Set AccessProtocol
+	pol, err := db.C.GetPool(ctx, vol.PoolId)
+	if err != nil {
+		msg := fmt.Sprintf("get pool failed in create volume attachment method: %v", err)
+		log.Error(msg)
+		return
+	}
+	var protocol = pol.Extras.IOConnectivity.AccessProtocol
```

Comment:
What happens if there is an error in getting protocol?

Reply:
AccessProtocol is an essential property for Pool and Profile, the system will choose pool to create a volume according to profile.it will not get error here because the pool must contain AccessProtocol